lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -3786,9 +3774,43 @@ ExpandMacroExpansionDeclRequest::evaluate(Evaluator &evaluator,
     return {};
   MED->setMacroRef(macro);
 
-  // Expand the macro.
-  SmallVector<Decl *, 2> expandedTemporary;
-  if (!expandFreestandingDeclarationMacro(MED, expandedTemporary))
+  auto roles = cast<MacroDecl>(macro.getDecl())->getMacroRoles();
+  // If it's not a declaration macro or a code item macro, it must have been
+  // parsed as an expression macro, and this decl is just its substitute decl.
+  // So there's no thing to be done here.
+  if (!roles.contains(MacroRole::Declaration))
     return {};
-  return ctx.AllocateCopy(expandedTemporary);
+
+  // Otherwise, we treat it as a declaration macro.
+  assert(roles.contains(MacroRole::Declaration));
+  return expandFreestandingMacro(MED);
+}
+
+ArrayRef<ASTNode>
+RecursivelyExpandMacroExpansionDeclRequest::evaluate(
```

Comment:
I'm really wary about introducing a request that does recursive expansion. Instead, if we `visitAuxiliaryDecls` in the right places throughout the frontend, that will facilitate recursive expansion through the request model anyway.

Reply:
Remove the request for now.